### PR TITLE
Refactor make_fx to better support hop subgraph tracing

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125363
* __->__ #125267

Code movement + minor rewrites. We extract the states of make_fx out and encapsulate them into a _MakefxTracer class. This allows us to create a new make_fx_tracer when tracing subgraphs, the actual logic for tracing subgraph is in the next diff.

Test Plan:
Existing tests.

### CODE REVIEW

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -1143,91 +1143,96 @@ def create_node(self, *args, **kwargs):
...
+        self.proxy_function_mode = nullcontext()
+        self.fx_tracer: Optional[Tracer] = None
+        # pre-autograd tracing uses per-dispatch-key modes,
+        # which requires the python dispatcher
+        if self.tracing_mode == "symbolic" or self.pre_dispatch:
+            self.python_dispatcher_mode = enable_python_dispatcher()
 
-    @functools.wraps(f)
-    def wrapped(*args):
+    # TODO: change to a context mananer _override_defaults_based_on_inputs(f, args)
```
Comment (ydwu4) [inline torch/fx/experimental/proxy_tensor.py:1177]: remove comment

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -1143,146 +1143,214 @@ def create_node(self, *args, **kwargs):
...
+        self.tracing_mode: str = tracing_mode
+        self._allow_non_fake_inputs: bool = _allow_non_fake_inputs
+        self.pre_dispatch: bool = pre_dispatch
+        self.record_module_stack: bool = record_module_stack
+        self._allow_fake_constant: bool = _allow_fake_constant
+        self._error_on_data_dependent_ops: bool = _error_on_data_dependent_ops
+
+        # All context managers and their states should be initialized before tracing based on the inputs
+        # and configurations.
+        # After tracing, thier states should be cleaned except for shape_env.
```
Comment (ydwu4) [inline torch/fx/experimental/proxy_tensor.py:1170]: shape_env is leaked out because of `t.shape_env = self.fake_tensor_mode.shape_env`

**Review by Chillee — COMMENTED** [review]
Can we get some tests of this? Or are you relying purely on the existing test suite?

**Comment by ydwu4** [comment]
yeah, this pr is supposed to be a code movement only change. Existing tests should suffice. Any suspicious places you want to test?

**Review by Chillee — APPROVED** [review]

**Comment by Chillee** [comment]
Ok looks fine to me.

**Pull request closed without merging**